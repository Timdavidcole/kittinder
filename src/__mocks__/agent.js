const Breeds = {
  all: jest.fn()
};

const Images = {
  get: jest.fn()
}

const Favourites = {
  post: jest.fn()
}

export default { Breeds, Images, Favourites };
