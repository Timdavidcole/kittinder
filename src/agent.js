import _superagent from "superagent";

const API_ROOT = "https://api.thecatapi.com/v1/";
const API_KEY = "a701ff7c-5713-41b4-b4e7-86671f2ce5a3";

const responseBody = res => res.body;

const requests = {
  get: (url, query) =>
    _superagent
      .get(`${API_ROOT}${url}`)
      .set("x-api-key", API_KEY)
      .query(query)
      .then(responseBody),
  post: (url, body) =>
    _superagent
      .post(`${API_ROOT}${url}`, body)
      .then(responseBody),
  put: (url, body) =>
    _superagent
      .put(`${API_ROOT}${url}`, body)
      .then(responseBody)
};

const Images = {
  all: (query) => requests.get(`images/search`, query),
  byAuthor: author =>
    requests.get(`/notices?author=${encodeURIComponent(author)}&limit=5`),
  favoritedBy: author =>
    requests.get(`/notices?favorited=${encodeURIComponent(author)}&limit=5`),
  get: slug => requests.get(`/notices/${slug}`),
  del: slug => requests.del(`/notices/${slug}`)
};

export default {Images};