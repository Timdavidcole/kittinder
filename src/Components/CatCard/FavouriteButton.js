import React from "react";
import agent from "../../agent";

class FavouriteButton extends React.Component {
  constructor(props) {
    super(props);
    this.state = { error: false }
    this.onClick = this.onClick.bind(this);
  }

  onClick() {
    agent.Favourites.post(this.props.imageId, this.props.userId)
      .then(data => null)
      .catch(error => {
        this.setState({
          error: true
        });
      });
  }

  render() {
    return (
      <button
        id={`favouriteButton${this.props.imageId}`}
        onClick={this.onClick}
      >
        Favourite
      </button>
    );
  }
}

export default FavouriteButton;
