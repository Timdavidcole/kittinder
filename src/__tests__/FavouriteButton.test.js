import FavouriteButton from "../Components/CatCard/FavouriteButton";
import React from "react";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom";
import { mount, shallow, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });

jest.mock("../agent");
import agent from "../agent";

beforeEach(() => {
  agent.Favourites.post.mockImplementation(() => {
    return Promise.resolve([{ message: "Completed" }]);
  });
});

test("checks Index snapshot", async () => {
  const favouriteButton = await renderer.create(<FavouriteButton />);

  let tree = favouriteButton.toJSON();
  expect(tree).toMatchSnapshot();
});

test("Onclick calls API successfully", async () => {
  const wrapper = await mount(<FavouriteButton userId={1} imageId={1} />);
  wrapper.find("button").simulate("click");
  expect(agent.Favourites.post).toBeCalled();
});

test("CatCard has correct breeds array from API call", async () => {
  agent.Favourites.post.mockImplementation(() => {
    return Promise.reject(new Error("fail"));
  });

  const wrapper = await shallow(<FavouriteButton />);
  wrapper.find("button").simulate("click");
  expect(agent.Favourites.post).toBeCalled();
  return Promise.reject(new Error("fail"))
    .then(() => {})
    .catch(() => {
      expect(wrapper.update().state("error")).toBe(true);
    });
});
