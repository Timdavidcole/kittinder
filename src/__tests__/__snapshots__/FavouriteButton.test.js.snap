// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checks Index snapshot 1`] = `
<button
  id="favouriteButtonundefined"
  onClick={[Function]}
>
  Favourite
</button>
`;
